lib/ansible/module_utils/facts/hardware/linux.py

```diff
@@ -329,6 +330,20 @@ def get_dmi_facts(self):
                         dmi_facts[k] = 'NA'
                 else:
                     dmi_facts[k] = 'NA'
+                if k == 'bios_version':
```

Comment:
Since this is non-dmi firmware info being collected, I think this doesn't need to be in the loop over DMI_DICT.items, and the ansible_architecture check could be outside. Something like:

``` python
        arch = collected_facts.get('ansible_architecture')

        if arch == 'sparc64':
            prom_dict = self._get_sparc64_prom_info()
            dmi_facts.update(prom_dict)

        if arch == 'ppc64':
            vernum_dict = self._get_ppc64_vernum()
            dmi_facts.update(vernum_dict)

        return dmi_facts

    def _parse_sparc64_cpuinfo_lines(self, cpuinfo_lines):
        prom_dict = {}

        for line in cpuinfo_lines:
            data = line.split(":", 1)
            key = data[0].strip()

            if key == "prom":
                prom_dict['bios_version'] = data[1].strip()

        return prom_dict

    def _get_sparc64_prom_info(self):
        prom_dict = {}

        # process firmware/bios version info for Linux on SPARC64 and PPC64 architectures
        if not os.access('/proc/cpuinfo', os.R_OK):
            return prom_dict

        cpuinfo_lines = get_file_lines('/proc/cpuinfo')

        cpuinfo_prom_dict = self._parse_sparc64_cpuinfo_lines(cpuinfo_lines)
        prom_dict.update(cpuinfo_prom_dict)

        return prom_dict

    def _get_ppc64_vernum(self):
        vernum_dict = {}

        if not os.access('/proc/device-tree/openprom/ibm,fw-vernum_encoded', os.R_OK):
            return vernum_dict

        vernum = get_file_content('/proc/device-tree/openprom/ibm,fw-vernum_encoded')

        if vernum is not None:
            vernum_dict['bios_version'] = vernum

        return vernum_dict
```